vocode/streaming/transcriber/deepgram_transcriber.py

```diff
@@ -17,55 +18,122 @@
+
+    class Config:
+        arbitrary_types_allowed = True
+
+
+ENCODING_MAPPING = {
+    AudioEncoding.LINEAR16: AudioEncodingModel(
+        name="linear16",
+        dtype=np.int16,
+        vad_chunk_size=1024,  # increase min size because I hit he error where vad chunk size was too small
+        vad_sampling_rate=16_000,  # recomended by google speech to text
+    ),
+    AudioEncoding.MULAW: AudioEncodingModel(
+        name="mulaw",
+        dtype=np.int8,
+        vad_chunk_size=256,
+        vad_sampling_rate=8_000,
+    ),
+}
+
+
+ConfigType = TypeVar("ConfigType", contravariant=True, bound=TranscriberConfig)
+
+
+class VadChunk(TypedDict):
+    chunk: bytes
+    timestamp: float
+    ignore: bool
+    is_silence: bool
+
+
+class VADWorker(AsyncWorker[VadChunk]):
 
-class VADWorker(AsyncWorker):
     def __init__(
         self,
-        input_queue: asyncio.Queue,
+        input_queue: asyncio.Queue[VadChunk],
         output_queue: asyncio.Queue,
-        transcriber: BaseAsyncTranscriber,
+        transcriber: BaseAsyncTranscriber[ConfigType],
         logger: Optional[logging.Logger] = None,
+        vad_threshold: float = 0.35,
+        volume_threshold: int = 700,
     ):
         super().__init__(input_queue, output_queue)
-        self.vad_buffer = b""
-        self.voiced_confidences = deque([0.0] * WINDOWS, maxlen=WINDOWS)
+        self.output_queue: asyncio.Queue[Transcription]
+        self.vad_buffer: bytes = b""
+        # # Constants for Silero VAD
+        self.WINDOWS = 3
+
+        self.transcriber = transcriber
+        self.encoding = ENCODING_MAPPING[
+            self.transcriber.transcriber_config.audio_encoding
+        ]
+
+        self.VAD_THRESHOLD = vad_threshold
+        self.VOLUME_THRESHOLD = volume_threshold
+        # Ensure minimum chunk size for Silero VAD (sr/chunk_size should be <= 31.25)
+        # min_chunk_size = math.ceil(self.transcriber.encoding.vad_sampling_rate / 31.25)
+        # self.encoding.vad_chunk_size = max(self.transcriber.encoding.chunk, min_chunk_size)
+
+        self.WINDOW_SIZE = (
+            self.WINDOWS
+            * self.encoding.vad_sampling_rate
+            // self.encoding.vad_chunk_size
+        )
+        self.voiced_confidences = deque([0.0] * self.WINDOWS, maxlen=self.WINDOWS)
+
+        self.PREFIX_PADDING_MS = 150  # Minimum speech duration to trigger detection
+        self.GRACE_PERIOD_MS = 300  # Grace period before resetting speech detection
         self.last_vad_output_time = 0
         self.speech_start_time = 0
         self.last_speech_time = 0
         self.logger = logger or logging.getLogger(__name__)
-        self.transcriber = transcriber
-
-    def int2float(self, sound):
-        abs_max = np.abs(sound).max()
+        # TODO if it's important, we can add more adaptive downsampling, but for now
+        # we are sticking with 16k for VAD linear16 and multiples of 16k
+        # for mulaw we do not downsample
+        assert (
```

Comment:
i don't think this assert is necessary since `encoding.vad_sampling_rate` is hardcoded and cannot be wrong